core/src/org/javarosa/xpath/expr/XPathFuncExpr.java

```diff
@@ -401,41 +411,59 @@ private static void checkArity(String name, int expectedArity, int providedArity
     }
 
     /**
-     * Given a handler registered to handle the function, try to coerce the function arguments into
-     * one of the prototypes defined by the handler. If no suitable prototype found, throw an eval
-     * exception. Otherwise, evaluate.
+     * Given a handler registered to handle the function, try to coerce the
+     * function arguments into one of the prototypes defined by the handler. If
+     * no suitable prototype found, throw an eval exception. Otherwise,
+     * evaluate.
      *
-     * Note that if the handler supports 'raw args', it will receive the full, unaltered argument
-     * list if no prototype matches. (this lets functions support variable-length argument lists)
+     * Note that if the handler supports 'raw args', it will receive the full,
+     * unaltered argument list if no prototype matches. (this lets functions
+     * support variable-length argument lists)
      *
      * @param handler
      * @param args
+     * @param ec
      * @return
      */
-    private static Object evalCustomFunction(IFunctionHandler handler, Object[] args, EvaluationContext ec) {
+    private static Object evalCustomFunction(IFunctionHandler handler, Object[] args,
+                                             EvaluationContext ec) {
         Vector prototypes = handler.getPrototypes();
         Enumeration e = prototypes.elements();
         Object[] typedArgs = null;
 
+        boolean argPrototypeArityMatch = false;
+        Class[] proto;
         while (typedArgs == null && e.hasMoreElements()) {
-            typedArgs = matchPrototype(args, (Class[])e.nextElement());
+            // try to coerce args into prototype, stopping on first success
+            proto = (Class[])e.nextElement();
+            typedArgs = matchPrototype(args, proto);
+            argPrototypeArityMatch = argPrototypeArityMatch ||

```

Comment:
If we don't find a prototype that is the same size as the arguments then it isn't a type mismatch, it is an arity error